src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
     columnName ~ columnType ^^ {
       case(name ~ ty) => Column(name, ty)
     }
+
+  def CategoricalParser = "Categorical" ~> "(" ~> """\d+""".r <~ ")" ^^ { n => MultinomialType(n.toInt) }
+  def BooleanParser = "Boolean" ^^ { s => BooleanType }
+  def dataType = CategoricalParser | BooleanParser
+
   def schemaDeclaration: Parser[SchemaDeclaration] =
-    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ^^ {
-      case (r ~ isQuery ~ "(" ~ attrs ~ ")") => {
+    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ~ opt(dataType) ^^ {
+      case (r ~ isQuery ~ "(" ~ attrs ~ ")" ~ vType) => {
         val vars = attrs.zipWithIndex map { case(x, i) => Variable(x.name, r, i) }
         var types = attrs map { case(x) => x.t }
-        SchemaDeclaration(Attribute(r, vars, types), (isQuery != None))
+        val variableType = vType match {
+          case None => if (isQuery != None) Some(BooleanType) else None
+          case Some(s) => Some(s)
+        }
+        SchemaDeclaration(Attribute(r, vars, types), (isQuery != None), variableType)
       }
     }
 
+  def operator = "||" | "+" | "-" | "*" | "/" | "&" | "::"
```

Comment:
How are we handling type casts?  I think it should be another case of expression because the rhs is not an expression but a type name unless we somehow treat types as first-class values, which sounds like permitting Russell's paradox.